### PR TITLE
Implement Object Selection

### PR DESCRIPTION
**Pull request by duncanspumpkin**
~~WiP will likely have to do a few other PR's before this one as there is a bit of structural stuff i want to improve prior.~~ complete

### CODE REVIEW

File: src/OpenLoco/src/Localisation/StringIds.h [hunk 1]
```diff
@@ -1648,9 +1648,12 @@ namespace OpenLoco::StringIds
     constexpr string_id buffer_2040 = 2040;
 
     constexpr string_id buffer_2042 = 2042;
-
+    constexpr string_id not_enough_space_for_graphics = 2043;
+    constexpr string_id too_many_objects_of_this_type_selected = 2044;
     constexpr string_id buffer_2045 = 2045;
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/Localisation/StringIds.h:1653]: technically these aren't really buffers (2045 and 2042) they are something else and comprise of an initial string followed by a buffer section.

File: src/OpenLoco/src/Localisation/LanguageFiles.cpp [hunk 2]
```diff
@@ -231,7 +231,7 @@ namespace OpenLoco::Localisation
     static bool stringIsBuffer(int id)
     {
         return id == StringIds::buffer_337 || id == StringIds::buffer_338 || id == StringIds::buffer_1250 || id == StringIds::preferred_currency_buffer || id == StringIds::buffer_1719
-            || id == StringIds::buffer_2039 || id == StringIds::buffer_2040 || id == StringIds::buffer_2042 || id == StringIds::buffer_2045;
+            || id == StringIds::buffer_2039 || id == StringIds::buffer_2040;
```
Comment (ZehMatt) [inline src/OpenLoco/src/Localisation/LanguageFiles.cpp:234]: At this point probably better to use a switch and let all of them fall through.

File: src/OpenLoco/src/Objects/Object.h [hunk 3]
```diff
@@ -45,6 +45,7 @@ namespace OpenLoco
         scenarioText,
     };
 
+    constexpr size_t kMaxObjectTypes = 34;
```
Comment (LeftofZen) [inline src/OpenLoco/src/Objects/Object.h:48]: This should be in S5/Limits.h

File: src/OpenLoco/src/Objects/ObjectIndex.cpp [hunk 4]
```diff
@@ -472,6 +485,238 @@ namespace OpenLoco::ObjectManager
...
+        // With also load objects it doesn't matter if they can't load
+        for (const auto& header : entry._alsoLoadObjects)
+        {
+            if ((mode & SelectObjectModes::selectAlsoLoads) != SelectObjectModes::none)
+            {
+                selectObjectFromIndexInternal(mode, true, header, objectFlags, selectionMetaData);
+            }
+        }
+
+        // When in don't load depenedent objects (required or also loads) mode then
```
Comment (LeftofZen) [inline src/OpenLoco/src/Objects/ObjectIndex.cpp:594]: typo on  'dependent', and grammatically the sentence doesn't make sense

File: src/OpenLoco/src/Objects/ObjectIndex.cpp [hunk 5]
```diff
@@ -472,6 +485,238 @@ namespace OpenLoco::ObjectManager
...
+        objectFlags[index] &= ~SelectedObjectsFlags::selected;
+        return true;
+    }
+
+    // 0x00473D1D
+    static bool selectObjectFromIndexInternal(SelectObjectModes mode, bool isRecursed, const ObjectHeader& objHeader, stdx::span<SelectedObjectsFlags> objectFlags, ObjectSelectionMeta& selectionMetaData)
+    {
+        if (!isRecursed)
+        {
+            // Vanilla had a few pointless no side effect loops here
```
Comment (LeftofZen) [inline src/OpenLoco/src/Objects/ObjectIndex.cpp:675]: maybe we just remove the whole `if` then?

File: src/OpenLoco/src/Windows/ObjectSelectionWindow.cpp [hunk 6]
```diff
@@ -1341,26 +1333,36 @@ namespace OpenLoco::Ui::Windows::ObjectSelectionWindow
...
 
         if ((selectionFlags[index] & ObjectManager::SelectedObjectsFlags::selected) == ObjectManager::SelectedObjectsFlags::none)
         {
-            bx |= (1 << 0); // Selection mode
+            mode = ObjectManager::SelectObjectModes::defaultSelect;
         }
 
-        bx |= 6;
+        ObjectManager::ObjectSelectionMeta meta{};
+        std::copy(std::begin(_112C1C5), std::end(_112C1C5), std::begin(meta.numSelectedObjects));
```
Comment (LeftofZen) [inline src/OpenLoco/src/Windows/ObjectSelectionWindow.cpp:1354]: Do you have a name for `_112C1C5`? I can make a PR to move it to a getter+setter

**Review by LeftofZen — COMMENTED** [review]
Few small comments, playtested it briefly and didn't find any issues with selecting objects or with limits for the number of objects

File: src/OpenLoco/src/Windows/ObjectSelectionWindow.cpp (hunk 6 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Windows/ObjectSelectionWindow.cpp:1354]: As far as i can work out it's just a temp global for the options window. I modify a few things related to it in the next few PRs so can address the name issue at the end of those

File: src/OpenLoco/src/Objects/Object.h (hunk 3 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/Objects/Object.h:48]: I'm a bit torn over moving it there. I like it here as its just below the object types. I'm not actually sure its even used in the s5 files as a limit its just used when iterating types.

**Review by AaronVanGeffen — APPROVED** [review]

**Pull request merged**